Improve the build system, add versioning, unit test framework

mjuric opened the pull request:
I apologize for a somewhat chunky PR, but will hopefully bring significant improvements (and enable future work, such as adding a testing suite and continuous integration). As it reworks much of the build system, I really hope a few of you can try it out to catch any major issues before merging. Most of this comes from LSST needs to integrate OpenOrb into the solar system processing pipelines.

What's in this PR:
1. Makefiles have been reworked to correctly record dependencies and rebuild code on an as-need basis.  Now it's also possible to run parallel builds, i.e.  something like:
   ```
   make -j4
   ```
   will work now, whereas it threw an error before. The default target `all` will build all programs from `main/` and store the binaries into `bin/`. It will also build the Python module, storing the library in `lib`. For backwards compatibility copies of binaries/modules are still kept in `main/` and `python/`.

   Because of how Fortran interacts with make (poorly), it's very difficult to correctly maintain a tree of Makefiles -- most have now been coalesced into a single one (in `build/Makefile`).

   All builds are now done in the `build/` directory. This makes it easier to make Fortran (various compilers) and `make` play well together. A beneficial side-effect is that source directories are not poluted by generated files.

   The detailed dependency rules are stored in build/make.depends files. This file is generated by running `make depends`, but we distribute it with the source code as it doesn't have to be updated unless module USE-age is changed or new files are added. See the documentation in `Makefile` for details.

   Tested to work with gfortran and ifort.

2. Version information is now inferred from git tags, and compiled into both the `oorb` binary (printable using `oorb --version`), and the `pyoorb` module (in `pyoorb.__version__`). Having proper versions helps with recording of provenance. Generated versions are PEP440 compliant. If the code is not being built in a git repository, whatever is stored in a file named `VERSION` will be used. This allows for tarball distributions. See the documentation in `build-tools/compute-version.sh` for the complete version construction logic.

3. Added a unit tests framework based on `pytest`. Now one can run:
   ```
    make test
   ```
   to run any pytest-compliant tests stored in `tests/`.  If ephemeris data are available, this will also run the `python/test.py` integration test file.

mjuric commented:
@mommermi -- could you give this PR a spin and a review? Would be good to have a second pair of eyes to make sure my changes didn't break `pyoorb`.

mommermi commented:
@mjuric: sure, will do!

mjuric commented on Makefile:
```diff
@@ -1,69 +1,98 @@
+#
+# OpenOrb Build System
+#
+# This file its companion in build/Makefile define build targets for all
```
Thanks, will fix.

mjuric commented on README.md:
````diff
@@ -29,58 +30,147 @@ of the resulting orbits.
 
-## Requirements ##
+```bash
+conda install openorb-data-de405
````
Thanks, good catch! Will fix.

N.b., the reason why I changed the docs to include the more complicated `-c defaults -c conda-forge` segment is because `conda-forge` has a tendency to replace a lot of packages that come with Anaconda with their own versions. This is something a casual user probably doesn't want.

mjuric commented:
One more question: you said this fails when running `sudo make install` -- does that mean that regular `make` successfully built `pyoorb`?

mommermi commented:
Sorry for the delay, had to do some more testing.

I am running this in a fresh Python 3.6 conda environment. Here's the info you wanted:

```
(oorb_install) mommermi@ulmo:~/Downloads/oorb/oorb$ uname -a
Linux ulmo 4.15.0-43-generic #46~16.04.1-Ubuntu SMP Fri Dec 7 13:31:08 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
(oorb_install) mommermi@ulmo:~/Downloads/oorb/oorb$ which python
/home/mommermi/software/anaconda3/envs/oorb_install/bin/python
(oorb_install) mommermi@ulmo:~/Downloads/oorb/oorb$ which gfortran
/usr/bin/gfortran
(oorb_install) mommermi@ulmo:~/Downloads/oorb/oorb$ which f2py
/home/mommermi/software/anaconda3/bin/f2py
(oorb_install) mommermi@ulmo:~/Downloads/oorb/oorb$ python -c 'import numpy; print(numpy.__file__)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'numpy'

```

Running this, I realized that `numpy` was not installed. So I installed it with pip (version 1.16.1), did a `make clean` and started from scratch. Interestingly, `make test` fails with a pytest failure if numpy is installed. I removed it again, did a `make clean` and then it ran through again up to `sudo make install` where it fails.

mommermi commented:
Unfortunately still the same problem...

```
(oorb-test) mommermi@ulmo:~/Downloads/oorb/oorb$ sudo make install
make[1]: Entering directory '/home/mommermi/Downloads/oorb/oorb/build'
analyzevars: charselector={'len': '11'} unhandled.analyzevars: charselector={'len': '3'} unhandled.In file included from ./_pyoorb_build/src.linux-x86_64-2.7/fortranobject.c:2:0:
./_pyoorb_build/src.linux-x86_64-2.7/fortranobject.h:7:20: fatal error: Python.h: No such file or directory
compilation terminated.
In file included from ./_pyoorb_build/src.linux-x86_64-2.7/fortranobject.c:2:0:
./_pyoorb_build/src.linux-x86_64-2.7/fortranobject.h:7:20: fatal error: Python.h: No such file or directory
compilation terminated.
error: Command "x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I./_pyoorb_build/src.linux-x86_64-2.7 -I/usr/lib/python2.7/dist-packages/numpy/core/include -I/usr/include/python2.7 -c ./_pyoorb_build/src.linux-x86_64-2.7/fortranobject.c -o ./_pyoorb_build/_pyoorb_build/src.linux-x86_64-2.7/fortranobject.o" failed with exit status 1
Makefile:98: recipe for target '../python/pyoorb.so' failed
make[1]: *** [../python/pyoorb.so] Error 1
make[1]: Leaving directory '/home/mommermi/Downloads/oorb/oorb/build'
Makefile:54: recipe for target 'all' failed
make: *** [all] Error 2
```

mjuric commented:
@mommermi and I met on Slack & sorted it out.

A quick summary of what was going on: I tried to be user-friendly and made `make install` first check if everything has been correctly built; that's the root cause of the problem. Details: When `sudo` runs make, it clears the environment. That will make the Anaconda python go away, and the system python re-appear on PATH (from make's point of view). System python is v2.7 -- different than 3.x that was used to build the module. Makefile detects that (as it should), and tries to rebuild, for `2.7`, with the system python. That's why you see all those `...-2.7` directories owned by root (which is why `make clean` can't get rid of them). However, system python doesn't have it's header files installed (that's probably `python-dev` package in Ubuntu), and that's why it fails on `Python.h`.

On top of this, there was also GNU vs. BSD incompatibility in how `find` was being called.

mommermi commented:
I can confirm that this PR works and that it doesn't break with any `pyoorb` functionality! Thanks, Mario!

mjuric commented:
Thanks @mommermi.

@mgranvik @fedorets -- I think it's now in your court if you want to try it out before merging. Happy to fix any remaining issues!

mjuric commented:
To give you a flavor of what this PR enables, yesterday I hooked up my fork of `oorb` to Azure Pipelines continuous integration service which now automatically builds it on a matrix of operating systems and Python versions every time I push a new patch into master.

Eight builds are made in total (macOS w. conda w. py27, 36, 37; Ubuntu w. conda w. py27, 36, 37; and Ubuntu with system Python 2.7 and 3.5) and it runs tests on all of them. The builds are done in parallel and typically take 3-4 minutes (for all of them). It's easy to add more systems and/or different compilers. It also builds pull requests, to verify they work correctly before being merged.

Demo:
* See the nice badge (== [![Build Status](https://dev.azure.com/mjuric/oorb/_apis/build/status/mjuric.oorb?branchName=master)](https://dev.azure.com/mjuric/oorb/_build/latest?definitionId=1?branchName=master)) in the README.md at https://github.com/mjuric/oorb
* Clicking on the badge gets you to https://dev.azure.com/mjuric/oorb/_build/latest?definitionId=1?branchName=master which lists the details of the recent builds. If the build failed, you can inspect the logs to figure out why.

Once this PR is merged, I'll open another one to add continuous integration like this to the main repository. I hope it will be useful in rapidly identifying any regressions or breakages.

mgranvik commented:
Mario,

this is great stuff. I will have a look at it next week when I'm back at work.

Mikael

On Feb 7, 2019 at 01:10, Mario Juric wrote:

> 
> To give you a flavor of what this PR enables, yesterday I hooked up my fork of oorb to Azure Pipelines continuous integration service
> which now automatically builds it on a matrix of operating systems and Python versions every time I push a new patch into master.
> 
> Eight builds are made in total (macOS w. conda w. py27, 36, 37; Ubuntu w. conda w. py27, 36, 37; and Ubuntu with system Python 2.7 and
> 3.5) and it runs tests on all of them. The builds are done in parallel and typically take 3-4 minutes (for all of them). It's easy to
> add more systems and/or different compilers. It also builds pull requests, to verify they work correctly before being merged.
> 
> Demo:
>
>  *  See the nice badge (== Build Status ) in the README.md at https://github.com/mjuric/oorb
>  *  Clicking on the badge gets you to https://dev.azure.com/mjuric/oorb/_build/latest?definitionId=1?branchName=master which lists the
>     details of the recent builds. If the build failed, you can inspect the logs to figure out why.
> 
> Once this PR is merged, I'll open another one to add continuous integration like this to the main repository. I hope it will be useful
> in rapidly identifying any regressions or breakages.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.[AMKDRNoejJzyTq_rh9UjxBrg97njTDGAks5vK28GgaJpZM4aiws2.gif]
> 
> 
>

mjuric commented:
Thanks Mikael!

PS: I added test coverage reporting this morning -- unsurprisingly, it's quite poor right now. Click on [![Test Coverage](https://codecov.io/gh/mjuric/oorb/branch/master/graph/badge.svg)](https://codecov.io/gh/mjuric/oorb) and browse; it allows you to see what the tests have been executing down to individual .f90 lines.

E.g., this report shows that (unsurprisingly) there's not much automated testing yet for the command line utilities, while on the `pyoorb` side, [oorb_ephemeris_basic](https://codecov.io/gh/mjuric/oorb/src/master/python/pyoorb.f90#L866) may be a good target to bring coverage closer to 100%.

mommermi commented:
Sign me up for the `pyoorb` tests!

mjuric commented:
Hi @mgranvik -- just a gentle ping on looking into this PR when you get a chance.

Siltala commented:
Hi all,

@mgranvik passed the duty of testing/reviewing this PR on our end to me.
These are some really welcome updates (and much of this is stuff I've wanted to do as well, but never got around to actually doing :) )

Anyway, with your branch 'make' fails for me on both gfortran version 8.2.0 and ifort (version 18.0.0) on my macOS system. I attached here the terminal output for both:
[make_gfortran.log](https://github.com/oorb/oorb/files/2884061/make_gfortran.log)
[make_ifort.log](https://github.com/oorb/oorb/files/2884064/make_ifort.log)

It looks like in both cases the problems arise from building pyoorb stuff (disclaimer: I've never used the python bindings or even built them before). I tested building pyoorb with the old build system which actually fails in the same way, so technically this isn't a regression. But whatever the issue may be, I would like to see it fixed before we start building pyoorb by default. Of course it could just be some odd misconfiguration on my end..

'make oorb' meanwhile seems to work fine on both compilers.

I also personally feel that, to begin with, by default we should only build the oorb binary itself to reduce the amount of required dependencies. To some extent this is a matter of preference so if @mgranvik does prefer to have pyoorb built by default, no worries.

The commit history itself could be cleaner - I'd suggest rebasing 995a88b and 37b5546 to the original commits. In a PR it's always more elegant to correct issues in commits by fixing the original commits themselves rather than correcting them in a new commit.

Finally, I would suggest installing the oorb binary into /usr/local/bin as opposed to /opt/oorb and, similarly, the python libraries into /usr/local/lib/pythonX.X/site-packages . This better conforms to unix standards and doesn't require the user to meddle with $PATH and $PYTHONPATH.

I'll let you know if anything else comes to mind!

mjuric commented:
Thanks @Siltala, all very good comments! Thanks!

I just returned from a longish trip and am dealing with a few other things, but I hope to update the PR by sometime next week.

Siltala commented:
As a little update on our end, I talked about this with Mikael yesterday and he's fine with everything being built by default. So don't worry about that particular comment.

mjuric commented:
Hi @Siltala & @mgranvik -- here's the cleaned up PR with the requested changes. Some notes:
* The code is now installed to `/usr/local` by default; I'm glad you asked for this as it saves some work on the `conda` packaging side (where I had to maintain a separate patch to have it install into `/usr/local`). Note that a consequence is that the data files are installed into `/usr/local/share/oorb`, to follow the usual LFS conventions. This shouldn't matter much to an end user, as `oorb` now knows to look for the data files in the `share/oorb` directory of wherever it was installed.
* I also made `pyoorb` not build by default -- I think you were right this is a more reasonable default, as not everyone has a full-blown Python dev environment set up. To build `pyoorb`, one now needs to configure with `./configure --with-pyoorb` (@mommermi, note the change).
* Finally, I added a few command line options to `./configure` to override names of the Python interpreter, `f2py`, etc... These are helpful when your Python is called `python3` (and your `f2py` is `f2py3`) as is the case on some Ubuntu installs.

PS: Also, I rebased it to the top of the master, so merging should work. The branch [passes CI](https://dev.azure.com/mjuric/oorb/_build/results?buildId=116).

mjuric commented:
@siltala Btw, I'd still like to understand why `pyoorb` won't build for you, maybe on a separate Issue?

My guess it it has to do with a missing dependency or something similar (for gfortran). Most of us use Anaconda on OSX which comes with most dependencies out-of-the-box.

For ifort there are a few pointers on the internet that this may be due to a missing SDK; tough to debug it for me because I only have a Linux license for ifort).

Siltala commented:
Looks good.
One thing I'd like to see is updating the build instructions 'for the impatient' in README.md to suggest installing into /usr/local rather than /opt (which also means there's no need to change the path)

Otherwise, this is now fine as far as I'm concerned.
I'll open a separate issue for my pyoorb building problems (which, again, do not block this PR as pyoorb won't build for me even with the old build system)

mjuric commented:
Ahh, missed that one when updating the README. Done.
